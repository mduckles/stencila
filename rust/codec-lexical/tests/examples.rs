use std::{fs::read_to_string, path::PathBuf};

use codec::{
    common::{eyre::Result, glob::glob},
    format::Format,
    EncodeOptions,
};

use codec_lexical::{decode, encode};
use common_dev::insta::{assert_json_snapshot, assert_snapshot, assert_yaml_snapshot};

/// Decode each example of a Lexical document and create JSON and Lexical snapshots
/// including snapshots for losses
#[test]
fn examples() -> Result<()> {
    let pattern = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests/examples")
        .canonicalize()?
        .to_string_lossy()
        .to_string()
        + "/**/*";

    for path in glob(&format!("{pattern}.lexical"))?
        .chain(glob(&format!("{pattern}.koenig"))?)
        .flatten()
    {
        let format = Format::from_path(&path);
        let ext = format.extension();

        let name = path
            .file_name()
            .map(|name| name.to_string_lossy())
            .and_then(|name| name.strip_suffix(&[".", &ext].concat()).map(String::from))
            .expect("should have format suffix");

        let original = read_to_string(path)?;

        let (article, info) = decode(&original, None)?;

        assert_json_snapshot!(format!("{name}.json"), article);
        assert_yaml_snapshot!(format!("{name}.decode.losses"), info.losses);

        let (lexical, info) = encode(
            &article,
            Some(EncodeOptions {
                format: Some(format.clone()),
                compact: Some(false),
                ..Default::default()
            }),
        )?;

        assert_snapshot!(format!("{name}.{ext}"), lexical);
        assert_yaml_snapshot!(format!("{name}.encode.losses"), info.losses);

        let (lexical, ..) = encode(
            &article,
            Some(EncodeOptions {
                format: Some(format),
                standalone: Some(true),
                compact: Some(false),
                ..Default::default()
            }),
        )?;
        assert_snapshot!(format!("{name}.standalone.{ext}"), lexical);
    }

    Ok(())
}
