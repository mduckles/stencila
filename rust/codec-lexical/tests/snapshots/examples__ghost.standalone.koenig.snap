---
source: rust/codec-lexical/tests/examples.rs
expression: redact(&lexical)
---
{
  "root": {
    "type": "root",
    "children": [
      {
        "type": "html",
        "html": "<stencila-paragraph id=xxx depth=0 ancestors=''><p slot=content><stencila-text id=xxx depth=1 ancestors=Paragraph>A document for testing Stencila&#x27;s integration with Ghost. This document is used as a test fixture, so be careful making changes because it is likely to affect those tests. </stencila-text></p></stencila-paragraph><stencila-paragraph id=xxx depth=0 ancestors=''><p slot=content><stencila-text id=xxx depth=1 ancestors=Paragraph>Paragraph with </stencila-text><stencila-strong id=xxx depth=1 ancestors=Paragraph><strong><stencila-text id=xxx depth=2 ancestors=Paragraph.Strong>bold</stencila-text></strong></stencila-strong><stencila-text id=xxx depth=1 ancestors=Paragraph> </stencila-text><stencila-emphasis id=xxx depth=1 ancestors=Paragraph><em><stencila-text id=xxx depth=2 ancestors=Paragraph.Emphasis>italic</stencila-text></em></stencila-emphasis><stencila-text id=xxx depth=1 ancestors=Paragraph> </stencila-text><stencila-emphasis id=xxx depth=1 ancestors=Paragraph><em><stencila-strong id=xxx depth=2 ancestors=Paragraph.Emphasis><strong><stencila-text id=xxx depth=3 ancestors=Paragraph.Emphasis.Strong>bold and italic</stencila-text></strong></stencila-strong></em></stencila-emphasis><stencila-text id=xxx depth=1 ancestors=Paragraph> </stencila-text><stencila-strikeout id=xxx depth=1 ancestors=Paragraph><s><stencila-text id=xxx depth=2 ancestors=Paragraph.Strikeout>strikethrough</stencila-text></s></stencila-strikeout><stencila-text id=xxx depth=1 ancestors=Paragraph> </stencila-text><stencila-underline id=xxx depth=1 ancestors=Paragraph><u><stencila-text id=xxx depth=2 ancestors=Paragraph.Underline>underline</stencila-text></u></stencila-underline><stencila-text id=xxx depth=1 ancestors=Paragraph> </stencila-text><stencila-code-inline id=xxx depth=1 ancestors=Paragraph code=code><code>code</code></stencila-code-inline><stencila-text id=xxx depth=1 ancestors=Paragraph> </stencila-text><stencila-subscript id=xxx depth=1 ancestors=Paragraph><sub><stencila-text id=xxx depth=2 ancestors=Paragraph.Subscript>subscript</stencila-text></sub></stencila-subscript><stencila-text id=xxx depth=1 ancestors=Paragraph> </stencila-text><stencila-superscript id=xxx depth=1 ancestors=Paragraph><sup><stencila-text id=xxx depth=2 ancestors=Paragraph.Superscript>superscript</stencila-text></sup></stencila-superscript><stencila-text id=xxx depth=1 ancestors=Paragraph> </stencila-text><stencila-text id=xxx depth=1 ancestors=Paragraph>highlight</stencila-text><stencila-text id=xxx depth=1 ancestors=Paragraph> and a </stencila-text><stencila-text id=xxx depth=1 ancestors=Paragraph>LOST Link</stencila-text></p></stencila-paragraph><stencila-heading id=xxx depth=0 ancestors='' level=1><h1 slot=content><stencila-text id=xxx depth=1 ancestors=Heading>Heading 1</stencila-text></h1></stencila-heading><stencila-heading id=xxx depth=0 ancestors='' level=2><h2 slot=content><stencila-text id=xxx depth=1 ancestors=Heading>Heading 2</stencila-text></h2></stencila-heading><stencila-heading id=xxx depth=0 ancestors='' level=3><h3 slot=content><stencila-text id=xxx depth=1 ancestors=Heading>Heading 3</stencila-text></h3></stencila-heading><stencila-paragraph id=xxx depth=0 ancestors=''></stencila-paragraph><stencila-quote-block id=xxx depth=0 ancestors=''><blockquote slot=content><stencila-paragraph id=xxx depth=1 ancestors=QuoteBlock><p slot=content><stencila-text id=xxx depth=2 ancestors=QuoteBlock.Paragraph>A quote</stencila-text></p></stencila-paragraph></blockquote></stencila-quote-block><stencila-paragraph id=xxx depth=0 ancestors=''><p slot=content><stencila-text id=xxx depth=1 ancestors=Paragraph>LOST Unknown (aside)</stencila-text></p></stencila-paragraph><stencila-paragraph id=xxx depth=0 ancestors=''><p slot=content><stencila-text id=xxx depth=1 ancestors=Paragraph>LOST List</stencila-text></p></stencila-paragraph><stencila-paragraph id=xxx depth=0 ancestors=''><p slot=content><stencila-text id=xxx depth=1 ancestors=Paragraph>LOST List</stencila-text></p></stencila-paragraph><stencila-thematic-break id=xxx depth=0 ancestors=''></stencila-thematic-break><stencila-image-object id=xxx depth=0 ancestors=''><img src=https://stencila.ghost.io/content/images/2025/01/stencila-128.png></img></stencila-image-object><stencila-code-block id=xxx depth=0 ancestors='' code='print(\"Hello world\")' programming-language=Python><pre><code>print(&quot;Hello world&quot;)</code></pre></stencila-code-block><stencila-table id=xxx depth=0 ancestors=''><table slot=rows><tr id=xxx depth=1 ancestors=Table><td id=xxx depth=2 ancestors=Table.TableRow><stencila-paragraph id=xxx depth=3 ancestors=Table.TableRow.TableCell><p slot=content><stencila-text id=xxx depth=4 ancestors=Table.TableRow.TableCell.Paragraph>A</stencila-text></p></stencila-paragraph></td><td id=xxx depth=2 ancestors=Table.TableRow><stencila-paragraph id=xxx depth=3 ancestors=Table.TableRow.TableCell><p slot=content><stencila-text id=xxx depth=4 ancestors=Table.TableRow.TableCell.Paragraph>B</stencila-text></p></stencila-paragraph></td></tr><tr id=xxx depth=1 ancestors=Table><td id=xxx depth=2 ancestors=Table.TableRow><stencila-paragraph id=xxx depth=3 ancestors=Table.TableRow.TableCell><p slot=content><stencila-text id=xxx depth=4 ancestors=Table.TableRow.TableCell.Paragraph>1</stencila-text></p></stencila-paragraph></td><td id=xxx depth=2 ancestors=Table.TableRow><stencila-paragraph id=xxx depth=3 ancestors=Table.TableRow.TableCell><p slot=content><stencila-text id=xxx depth=4 ancestors=Table.TableRow.TableCell.Paragraph>2</stencila-text></p></stencila-paragraph></td></tr></table></stencila-table><stencila-raw-block id=xxx depth=0 ancestors='' format=html content='&lt;span&gt;HTML card&lt;/span&gt;'><div slot=content><span>HTML card</span></div></stencila-raw-block>"
      }
    ]
  }
}
